src/ast/passes/semantic_analyser.cpp

```diff
@@ -2803,6 +2820,44 @@ void SemanticAnalyser::visit(AssignConfigVarStatement &assignment)
   Visit(assignment.expr);
 }
 
+void SemanticAnalyser::visit(VarDeclStatement &decl)
+{
+  std::string var_ident = decl.var->ident;
+
+  // This should be the first time we're seeing this variable
+  auto decl_search = declared_variables_[scope_].find(var_ident);
+  if (decl_search != declared_variables_[scope_].end()) {
+    if (!areLocationsEqual(decl_search->second, decl.loc)) {
+      // Note this will change with proper block/lexical scoping
+      LOG(ERROR, decl.loc, err_)
+          << "Variable " << var_ident << " was already declared.";
+    }
+  }
+
+  auto search = variable_type_[scope_].find(var_ident);
+  if (search != variable_type_[scope_].end()) {
+    if (!declared_variables_[scope_].contains(var_ident)) {
+      LOG(ERROR, decl.loc, err_)
+          << "Variable declarations need to occur before variable usage or "
+             "assignment. Variable: "
+          << var_ident;
+    } else {
+      decl.var->type = search->second.type;
```

Comment:
IIUC, this is to assign type to untyped declarations after we have visited the assignment. Shouldn't we check that `decl.var->type` is not already set to a different (incompatible) type? Or that cannot happen?

Anyways, a comment here would help a lot, it took me a while to figure out the purpose of this line.

Reply:
> Shouldn't we check that decl.var->type is not already set to a different (incompatible) type? Or that cannot happen?

That can't happen because if we haven't seen this variable before (which we also validate in this function) then when the declaration happens we insert into the, newly renamed, `variable_type_` map. Then when var assignment  happens it will check against the type in that map and fail if they're not same (unless the store type is a None type, in which case it updates it). But yes, will add a comment :)